futures-util/src/future/flatten.rs

```diff
@@ -11,14 +12,13 @@ use super::chain::Chain;
 ///
 /// This is created by the `Future::flatten` method.
 #[must_use = "futures do nothing unless polled"]
-pub struct Flatten<A> where A: Future, A::Item: IntoFuture {
-    state: Chain<A, <A::Item as IntoFuture>::Future, ()>,
+pub struct Flatten<A> where A: Future, A::Output: Future {
```

Comment:
Should we also have Result-y versions of this and `FlattenStream`? Or maybe we just make `impl<T> Future for Result<T, E> where T: Future` (and the same for `Stream`)?